app/controllers/users_controller.rb

```diff
@@ -86,18 +86,13 @@ def restrict_after_lock
   end
 
   def updated_params
-    batch = nil
-
-    if current_user.batch_id.nil? && form_params[:city_id] # rubocop:disable Style/IfUnlessModifier
-      batch = Batch.find_or_create_by(number: nil, city_id: form_params[:city_id])
-    end
-
     {
       accepted_coc: form_params[:accepted_coc],
       aoc_id: form_params[:aoc_id],
       entered_hardcore: form_params[:entered_hardcore],
       username: form_params[:username],
-      batch_id: batch&.id,
+      batch_id: Batch.find_by(number: form_params[:batch_number])&.id,
+      city_id: form_params[:city_id],
       referrer: User.find_by_referral_code(form_params[:referrer_code])
     }.compact
```

Comment:
Same as previous comment, the batch should never be updated by the user event if he has none.